packages/core-data/src/actions.js

```diff
@@ -543,6 +558,34 @@ export function* saveEntityRecord(
 	}
 }
 
+export function* __experimentalBatchSaveEntityRecords( spec ) {
```

Comment:
This doesn't handle deletes. Not sure how the API for that should look.

Reply:
Hmm, yeah, I think whatever executes the batch shouldn't have to be aware of what's in the batch.

I was thinking that creation of the batch would happen in `saveEditedWidgetArea`:

```jsx
export function* saveWidgetArea( widgetAreaId ) {
    // ...
    // ... experimentalCreateBatch is a version of `createBatch` preconfigured with our batch 'processor'.
    const batch = __experimentalCreateBatch();
    // ...
    dispatch( 'core' ).saveEditedEntityRecord( 'root', 'widget', widgetId, { batch } );
    // ...
    // batch.run implicitly `waitsForSize`.
    batch.run();
}
```

Is it that we need to know the full batch size before calling `batch.run()`?

Potentially the batch could be made to be internally aware of that:
```jsx
// createItem increments the batchers internal count
const batchItem = batch.createItem();
// batchItem can add an item to a batch via its own method, e.g. `batchItem.addRequest`.
dispatch( 'core' ).saveEditedEntityRecord( 'root', 'widget', widgetId, { batchItem } );
```

The alternative is that the batch spec knows which method to call:
```jsx
batchSpec.push( {
    action: dispatch( 'core' ).saveEditedEntityRecord,
    args: [ 'root', 'widget', widgetId ],
} );
```

Reply:
> I was thinking that creation of the batch would happen in saveEditedWidgetArea:

I'm not sure I understand why you  think the batch should be created out of core-data. Having the way it's now n the PR has the advantage of being an implementation detail and can make this useful for all screens (multi-entity  saving  flow).

Reply:
> I'm not sure I understand why you think the batch should be created out of core-data. Having the way it's now n the PR has the advantage of being an implementation detail and can make this useful for all screens (multi-entity saving flow).

The way I'm describing, the batch processor would still be part of core data, so that indeed it can be used by all consumers of that package. We can have a generic `createBatch( processor )` function that can be used in any project. Core data can export a version of that that's preconfigured with its processor that uses the batch endpoint. That way it wouldn't have to be re-configured when imported.

What I was describing is that I think the way the batch is assembled could be different. At the moment `__experimentalBatchSaveEntityRecords` has to know how to call `saveEditedEntityRecord`, `saveEntityRecord`, `deleteEntityRecord` and so on, but I don't think it needs to, and I don't think this approach is scaleable. The processor should have just enough information to run the batch, but shouldn't need knowledge of our data API.

Reply:
> I don't think it needs to, and I don't think this approach is scaleable

Why do you think it's not scalable enough?  For me `__experimentalBatchSaveEntityRecords` could be named just  `__experimentalSaveEntityRecords` and the batch be just an implementation detail external users of core-data might not even need to know about.

Reply:
What if we want a batch that's not just for saving? I think that was the original main use case on the navigation screen where we want to be able to create, edit and delete menu items in a particular order.

Reply:
> What if we want a batch that's not just for saving?

That's a good point, but I don't see how we could have a nice API for that while keeping the processor inside core-data.

Reply:
Good ideas here! Seems there's three criteria:

1. The caller needs to be able to do different types of requests at once. (Save, delete, ...possibly more?)
2. The caller shouldn't have to care about what a "batch" is as that's implementation detail-ey.
3. The API should be "nice" 😛

Here's where I ended up after some iteration this afternoon:


```js
const [ newRecord, savedRecord, deletedRecord ] = await wp.data.dispatch( 'core' ).performBatch( [
	( { saveEntityRecord } ) => saveEntityRecord( 'root', 'widget', widget ),
	( { saveEditedEntityRecord } ) => saveEditedEntityRecord( 'root', 'widget', 123 ),
	( { deleteEntityRecord } ) => deleteEntityRecord( 'root', 'widget', 456 ),
] )
// Can use getLastEntitySaveError() or getLastEntityDeleteError() here if needed.
```

094e6e3363 implements this. I'm pretty happy with it as it feels to me like a pretty natural extension of the existing `core-data` API. Let me know what you think though.

Reply:
Seing this make me wonder if it could be a "data" API

```
wp.data.batch( [
	( dispatch ) => dispatch('core').saveEntityRecord( 'root', 'widget', widget ),
	( dispatch ) => dispatch('core').saveEditedEntityRecord( 'root', 'widget', 123 ),
	( dispatch ) => dispatch('core').deleteEntityRecord( 'root', 'widget', 456 ),
] )
```

In an ideal world this would only trigger a single "subscribe" call at the end of the call.
It could also serve as a performance  optimization: for instance RichText does both onChange and onSelectionChange on the same tick. good candidate for a batch.

That said, it does seem a bit ambitious to start with so maybe we can just go with what we have as "experimental"

Reply:
That would be really cool 🙂